### PR TITLE
Run Swift file length budget in pre-commit

### PR DESCRIPTION
**Pull request by lawrencecchen**
## Summary
- Add a tracked `.githooks/pre-commit` hook that runs the Swift file length budget against staged Swift blobs.
- Add fast staged/path-limited modes to the checker while keeping the full CI scan path.
- Install the tracked hook path from `scripts/setup.sh` and cover the hook in workflow guard tests.

## Testing
- `./tests/test_ci_swift_file_length_budget.sh`
- `./tests/test_ci_pre_commit_hooks.sh`
- `python3 -m py_compile scripts/swift_file_length_budget.py`
- `.githooks/pre-commit`

## Issues
- Plain-text task: run file length checker as pre-commit hook and make it faster

---

> [!NOTE]
> **Low Risk**
> Low risk: changes are limited to developer tooling (git hook + budget checker) and accompanying tests, with no runtime production impact. Main risk is false positives/negatives in the new index-based scanning paths affecting developer workflows.
> 
> **Overview**
> Adds a tracked `.githooks/pre-commit` hook that enforces the Swift file length budget on *staged* Swift blobs, and additionally re-checks the entire repo from the git index when the budget file itself is staged.
> 
> Extends `scripts/swift_file_length_budget.py` with an `--index` mode, more robust root/path normalization (including `--roots .`), clearer argument validation, and a stricter error when `--paths` matches no owned Swift files. Updates CI shell tests to cover the new pre-commit behavior and the new `--roots`/missing-path cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3ad9bb0901b44682f45f7cc469f41fcab387b5c3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
<a href="https://app.blacksmith.sh/manaflow-ai/codesmith/cmux/pr/4664"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1782130108&installation_id=133855650&pr_number=4664&repository=manaflow-ai%2Fcmux&return_to=https%3A%2F%2Fgithub.com%2Fmanaflow-ai%2Fcmux%2Fpull%2F4664&signature=ec22a5921941934672c21e5eb2cbadcad37e4ad8a336e0809229a4b3fe12b4fb"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Runs the Swift file length budget as a Git pre-commit hook. It checks staged Swift blobs and the staged budget, and runs a full index scan when the budget file is staged, while keeping fast staged/path-limited modes and the full CI scan.

- New Features
  - Tracked `.githooks/pre-commit` runs `scripts/swift_file_length_budget.py --staged` on staged Swift blobs, reads the staged budget, and if the budget file is staged runs `--index` against the staged snapshot.
  - Added `--staged`, `--index`, and `--paths` modes with compare-on-checked-paths only and clearer errors for unmatched paths or invalid flag combos.
  - Faster binary newline counting and deterministic directory walking.
  - CI adds pre-commit hook tests and path-limited checks; `scripts/setup.sh` sets `core.hooksPath=.githooks`.

- Migration
  - Run `scripts/setup.sh` once to install the Git hooks.

<sup>Written for commit 3ad9bb0901b44682f45f7cc469f41fcab387b5c3. Summary will update on new commits. <a href="https://cubic.dev/pr/manaflow-ai/cmux/pull/4664?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **New Features**
  * Pre-commit hook now enforces Swift file-length budgets and respects staged/INDEX/paths selection for validation.
  * CLI expanded to accept --paths, --staged and --index modes for targeted budget checks.

* **Chores**
  * Development setup configures Git hooks automatically.
  * CI workflow runs the pre-commit hook validation as part of guard tests.

* **Tests**
  * Added/expanded tests covering hook behavior, staged/untracked Swift files, and CLI path-flag validation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/manaflow-ai/cmux/pull/4664?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: scripts/swift_file_length_budget.py [hunk 1]
```diff
@@ -183,15 +294,47 @@ def main(argv: list[str]) -> int:
         action="store_true",
         help="write the current file lengths as the budget instead of checking",
     )
+    parser.add_argument(
+        "--paths",
+        nargs="*",
+        help="optional repo-relative or absolute paths to check instead of scanning every root",
+    )
```
[the remark this answers is not in the document]
Comment (lawrencecchen) [reply scripts/swift_file_length_budget.py:360]: Fixed by changing --paths to require at least one argument and adding a guard test for empty --paths.

*— Claude Code*

File: scripts/swift_file_length_budget.py [hunk 2]
```diff
@@ -183,18 +314,53 @@ def main(argv: list[str]) -> int:
...
+        file_lengths = collect_file_lengths(repo_root, roots)
+    else:
+        file_lengths = collect_file_lengths_for_paths(repo_root, roots, args.paths)
+        checked_paths = set(file_lengths)
     actual = tracked_file_lengths(file_lengths, args.threshold)
     print_file_summary("All scanned cmux-owned Swift files", file_lengths)
     print_file_summary(f"Tracked Swift files >= {args.threshold} lines", actual)
+    if checked_paths is not None and not checked_paths:
+        print("Swift file length budget respected.")
+        return 0
```
[the remark this answers is not in the document]
Comment (lawrencecchen) [reply scripts/swift_file_length_budget.py:425]: Fixed by making explicit --paths invocations fail when they do not resolve to any cmux-owned Swift file, while preserving the clean no-op for staged commits with no Swift files. Added shell coverage for both cases.

*— Claude Code*
